### PR TITLE
[Feature][Connector-V2][Teradata] Add Teradata Source And Sink Connector 

### PR DESCRIPTION
**Pull request by FWLamb**
## Purpose of this pull request

Refer to https://github.com/apache/incubator-seatunnel/issues/3018
Add Teradata Source And Sink Connector.

Check list

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [x] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

**Comment by FWLamb** [comment]
725ef347a4c95beaae5fcae522e5092

**Comment by FWLamb** [comment]
image
image

**Comment by TaoZex** [comment]
Do we need to use jdbc dialect? @hailin0

**Comment by hailin0** [comment]
> Do we need to use jdbc dialect? @hailin0

agree you @TaoZex 

@FWLamb  reference
https://github.com/apache/incubator-seatunnel/pull/2410

**Comment by FWLamb** [comment]
534f8447470f0e2cf630991ba0009ae

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/connection/SimpleJdbcConnectionProvider.java [hunk 1]
```diff
@@ -121,24 +121,29 @@ public Connection getOrEstablishConnection()
...
             throw new SQLException(
-                "No suitable driver found for " + jdbcOptions.getUrl(), "08001");
+                "No suitable driver found for " + jdbcOptionsUrl, "08001");
         }
 
-        //Auto commit is used by default
-        connection.setAutoCommit(true);
-
+        // Auto commit is used by default, Teradata jdbc type fastload must be set false.
+        connection.setAutoCommit(!(jdbcOptionsUrl.toLowerCase().contains("jdbc:teradata:") & jdbcOptionsUrl.toLowerCase().contains("type=fastload")));
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/connection/SimpleJdbcConnectionProvider.java]: cc @ic4y

Comment (ic4y) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/connection/SimpleJdbcConnectionProvider.java]: Waiting for the #3451 pr to be merged, there is no need to modify it.

**Review by TaoZex — DISMISSED** [review]
LGTM

**Comment by FWLamb** [comment]
image

**Comment by FWLamb** [comment]
@ic4y @hailin0 @EricJoy2048 @TaoZex  PTAL.

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/utils/JdbcUtils.java [hunk 2]
```diff
@@ -39,21 +47,31 @@ public class JdbcUtils {
...
+        throws SQLException {
         if (typesArray != null && typesArray.length > 0 && typesArray.length != row.getFields().length) {
             LOG.warn(
-                    "Column SQL types array doesn't match arity of passed Row! Check the passed array...");
+                "Column SQL types array doesn't match arity of passed Row! Check the passed array...");
         }
         if (typesArray == null) {
             // no types provided
             for (int index = 0; index < row.getFields().length; index++) {
-                upload.setObject(index + 1, row.getFields()[index]);
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/utils/JdbcUtils.java]: In #3444, there is a better way to do it. It is recommended to wait for #3444 to be merged first

Comment (FWLamb) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/utils/JdbcUtils.java]: done, PTAL.

Comment (FWLamb) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/utils/JdbcUtils.java]: done. PTAL

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcTeradataIT.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    };
+
+    private Connection connection;
+
+    @TestTemplate
+    public void testTeradata(TestContainer container) throws Exception {
+        container.executeExtraCommands(extendedFactory);
+        Container.ExecResult execResult = container.executeJob("/jdbc_teradata_source_and_sink.conf");
+        Assertions.assertEquals(0, execResult.getExitCode());
+        clearSinkTable();
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcTeradataIT.java:56]: @FWLamb 
Please reference to `JdbcOracledbIT` validation data input and output


https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOracledbIT.java

Comment (FWLamb) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcTeradataIT.java:56]: Teradata does not have a Docker image, and it seems that it cannot be made by itself.

**Review by ic4y — APPROVED** [review]
LGTM

**Review by EricJoy2048 — APPROVED** [review]

**Pull request merged**